### PR TITLE
[TRB-44974] Dynamic Configuration for Node Scaling in Kubeturbo

### PR DESCRIPTION
**Pull request by libai98**
# Intent

_set the min/max node information in the configmap of kubeturbo to be auto loaded for kubeturbo_

# Background
For customers looking to gain insights into cluster capacity for running their workloads, we offer provision and suspend actions as a means to understand real-time demand. By utilizing our analysis, customers can accurately determine the number of nodes required, as opposed to relying solely on thresholds used with cluster autoscalers. Our customers seek to align these actions with their specific definitions of minimum and maximum node numbers, and they require Turbo to discover these definitions and allow them to define policies regarding node suspension and provisioning.

wiki: https://ibm-turbo-wiki.fyre.ibm.com/en/CON/AE/Advanced-Engineering/Cloud-Native-Control/new-page

# Testing

Edit config map yaml with different cluster minNodes and maxNodes bvalues and apply the change with command 
vi configMap.yaml
oc -n turbo apply -f configMap.yaml

oc -n turbo  logs -f kubeturbo-c9b7c9f4c-8tbks  | grep -i "Cluster clus" -A1 -B1
```
********MinNodes = 10, maxNodes not specified
I0808 02:52:24.146801       1 kubeturbo_builder.go:1038] Start watching the autoreload config file /etc/kubeturbo/turbo-autoreload.config
I0808 02:52:24.146823       1 kubeturbo_builder.go:1072] Cluster cluster.minNodes  current value is 1
I0808 02:52:24.148957       1 kubeturbo_builder.go:1088] Cluster cluster.minNodes changed from 1 to 10
I0808 02:52:24.148966       1 kubeturbo_builder.go:1072] Cluster cluster.maxNodes  current value is 1000
I0808 02:52:24.148973       1 kubeturbo_builder.go:1076] Empty value for cluster.maxNodes in the configuration, using default value 1000
--

********MinNodes =15, maxNodes= 180
I0808 02:52:38.300563       1 api_client.go:320] Successfully updated target via API service.
I0808 02:57:42.787640       1 kubeturbo_builder.go:1072] Cluster cluster.minNodes  current value is 10
I0808 02:57:42.787665       1 kubeturbo_builder.go:1088] Cluster cluster.minNodes changed from 10 to 15
I0808 02:57:42.787670       1 kubeturbo_builder.go:1072] Cluster cluster.maxNodes  current value is 1000
I0808 02:57:42.787676       1 kubeturbo_builder.go:1088] Cluster cluster.maxNodes changed from 1000 to 180
I0808 02:58:41.510006       1 remote_mediation_client.go:216] [handleServerMessages][ServerRequestEndpoint] : received message with request type Discovery.
--

********MinNodes =-15, maxNodes= 185
I0808 02:59:00.796572       1 remote_mediation_client.go:363] Discovery has finished for 4217:FULL
I0808 02:59:11.622246       1 kubeturbo_builder.go:1072] Cluster cluster.minNodes  current value is 15
E0808 02:59:11.622273       1 kubeturbo_builder.go:1083] Invalid cluster.minNodes value: -15 specified, using default value 1
I0808 02:59:11.622493       1 kubeturbo_builder.go:1072] Cluster cluster.maxNodes  current value is 180
I0808 02:59:11.622516       1 kubeturbo_builder.go:1088] Cluster cluster.maxNodes changed from 180 to 185


```
# Checklist

These are the items that must be done by the developer and by reviewers before the change is ready to merge. Please ~~strikeout~~ any items that are not applicable, but don't delete them

- [ ] Developer Checks
    - [X] Full build with unit tests and fmt and vet checks
    - [X] Unit tests added / updated
    - [X] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Integration tests added / updated
    - [X] Manual testing done (and described)
    - [ ] Product sweep run and passed
    - [X] Developer wiki updated (and linked to this description)
- [ ] Reviewer Checks
    - [ ] Merge request description clear and understandable
    - [ ] Developer checklist items complete
    - [ ] Functional code review (how is the code written)
    - [ ] Architectural review (does the code try to do the right thing, in the right way)
    - [ ] Defensive coding (incoming data checked / sanitized, exceptions logged, clear error messages)
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Security review checklist complete.

# Audience

_(@ mention any `review/...` groups or people that should be aware of this merge request)_

### CODE REVIEW

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 1]
```diff
@@ -1030,28 +1031,60 @@ func WatchConfigMap() {
...
 		newLoggingLevel := viper.GetString("logging.level")
 		currentLoggingLevel := pflag.Lookup("v").Value.String()
-		if newLoggingLevel != currentLoggingLevel {
-			if newLogVInt, err := strconv.Atoi(newLoggingLevel); err != nil || newLogVInt < 0 {
-				glog.Errorf("Invalid log verbosity %v in the autoreload config file", newLoggingLevel)
-			} else {
-				err := pflag.Lookup("v").Value.Set(newLoggingLevel)
-				if err != nil {
-					glog.Errorf("Can't apply the new logging level setting due to the error:%v", err)
+		if newLoggingLevel != "" {
```
Comment (amd-ibm) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1041]: nit: you could AND these IFs

**Review by amd-ibm — APPROVED** [review]

**Comment by libai98** [comment]
Some new test logs related to the configmap syntax change:
 ```
oc -n turbo  logs -f kubeturbo-587654cdc5-mdszh  | grep -i "Cluster nodePool" -A1 -B1
I0809 14:33:15.150037       1 kubeturbo_builder.go:1037] Start watching the autoreload config file /etc/kubeturbo/turbo-autoreload.config
I0809 14:33:15.152271       1 kubeturbo_builder.go:1070] Cluster nodePoolSize.min current value is 1
I0809 14:33:15.152279       1 kubeturbo_builder.go:1074] Empty value for nodePoolSize.min in the configuration, using default value 1
I0809 14:33:15.152282       1 kubeturbo_builder.go:1070] Cluster nodePoolSize.max current value is 1000
I0809 14:33:15.152286       1 kubeturbo_builder.go:1074] Empty value for nodePoolSize.max in the configuration, using default value 1000
--
I0809 14:36:24.404329       1 remote_mediation_client.go:363] Discovery has finished for 532:FULL
I0809 14:36:56.676214       1 kubeturbo_builder.go:1070] Cluster nodePoolSize.min current value is 1
I0809 14:36:56.676246       1 kubeturbo_builder.go:1087] Cluster nodePoolSize.min changed from 1 to 25
I0809 14:36:56.676253       1 kubeturbo_builder.go:1070] Cluster nodePoolSize.max current value is 1000
I0809 14:36:56.676261       1 kubeturbo_builder.go:1087] Cluster nodePoolSize.max changed from 1000 to 1050
I0809 14:38:17.677010       1 kubeturbo_builder.go:1070] Cluster nodePoolSize.min current value is 25
I0809 14:38:17.677039       1 kubeturbo_builder.go:1074] Empty value for nodePoolSize.min in the configuration, using default value 1
I0809 14:38:17.677044       1 kubeturbo_builder.go:1070] Cluster nodePoolSize.max current value is 1050
I0809 14:38:17.677053       1 kubeturbo_builder.go:1087] Cluster nodePoolSize.max changed from 1050 to 123
```

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 2]
```diff
@@ -1044,14 +1048,43 @@ func WatchConfigMap() {
 				} else {
 					glog.V(1).Infof("Logging level is changed from %v to %v", currentLoggingLevel, newLoggingLevel)
 				}
-
 			}
 		}
+
```
Comment (kevinwangcn) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1053]: I don't think you need to pass `viper.GetString`  here, why don't you call it directly in the function `updateNodePoolConfig`？

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 3]
```diff
@@ -1030,11 +1031,14 @@ func WatchConfigMap() {
...
+	currentMinNodes := cluster.DefaultMinNodePoolSize
+	currentMaxNodes := cluster.DefaultMaxNodePoolSize
+
 	glog.V(1).Infof("Start watching the autoreload config file %s/%s", autoReloadConfigFilePath, autoReloadConfigFileName)
-	updateConfig := func() {
+	updateConfigClosure := func() {
 		newLoggingLevel := viper.GetString("logging.level")
 		currentLoggingLevel := pflag.Lookup("v").Value.String()
-		if newLoggingLevel != currentLoggingLevel {
+		if newLoggingLevel != "" && newLoggingLevel != currentLoggingLevel {
```
Comment (kevinwangcn) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1041]: This looks good to me. Thanks for making this change.

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 4]
```diff
@@ -1044,14 +1048,43 @@ func WatchConfigMap() {
...
+
+// updateNodePoolConfig updates the value of a configuration setting for an integer property.
+// If the configuration value is empty, it uses the provided defaultValue.
+// If the configuration value is not a valid integer or is negative, it uses the defaultValue and logs an error.
+// If the configuration value is different from the currentValue, it updates the currentValue and logs the change.
+func updateNodePoolConfig(configKey string, currentValue *int, defaultValue int, getKeyStringVal func(string) string) {
+	glog.V(1).Infof("Cluster %s current value is %v ", configKey, *currentValue)
+	newValueStr := getKeyStringVal(configKey)
+
+	if newValueStr == "" {
```
Comment (kevinwangcn) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1073]: Suppose the user ONLY defined `logging.level` in the configMap,  whenever the user updates the `logging.level`, this `("Empty value for %s in the configuration, using default value %d", configKey, defaultValue)` will be printed out according to the current logic.   You may need to make the similar change as what you did for the logging level.

Comment (libai98) [reply cmd/kubeturbo/app/kubeturbo_builder.go:1073]: I think that is what we intended to do. If we do not specify the value for min and max of the node group, regardless of what the current values for min and max, we will revert that to the default.

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 5]
```diff
@@ -1030,11 +1031,14 @@ func WatchConfigMap() {
...
 	glog.V(1).Infof("Start watching the autoreload config file %s/%s", autoReloadConfigFilePath, autoReloadConfigFileName)
-	updateConfig := func() {
+	updateConfigClosure := func() {
 		newLoggingLevel := viper.GetString("logging.level")
 		currentLoggingLevel := pflag.Lookup("v").Value.String()
-		if newLoggingLevel != currentLoggingLevel {
+		if newLoggingLevel != "" && newLoggingLevel != currentLoggingLevel {
 			if newLogVInt, err := strconv.Atoi(newLoggingLevel); err != nil || newLogVInt < 0 {
 				glog.Errorf("Invalid log verbosity %v in the autoreload config file", newLoggingLevel)
 			} else {
```
Comment (ading1977) [inline cmd/kubeturbo/app/kubeturbo_builder.go:1044]: Shall we extract the logging level change into a function, e.g., `updateLoggingLevel`?

File: cmd/kubeturbo/app/kubeturbo_builder.go (hunk 4 above)
Comment (ading1977) [reply cmd/kubeturbo/app/kubeturbo_builder.go:1073]: I think we should only print out message when there is actually an effective change. In most of the cases, there won't be settings for min/max node groups in the configMaps, then every time user changes a logging level, there will be a message printed out for min/max nodegroup which is not necessary and could actually be confusing.

File: cmd/kubeturbo/app/kubeturbo_builder.go (hunk 2 above)
Comment (kevinwangcn) [reply cmd/kubeturbo/app/kubeturbo_builder.go:1053]: I see your point

**Review by kevinwangcn — APPROVED** [review]

**Review by ading1977 — COMMENTED** [review]
There should also be changes to the operator, right>

File: cmd/kubeturbo/app/kubeturbo_builder.go (hunk 4 above)
Comment (libai98) [reply cmd/kubeturbo/app/kubeturbo_builder.go:1073]: I think the solution is:
1) If the current min/max value is not default, we will print out the message of reverting to default if min/max were not specified in the configmap with message printed out
2) If the current min/max value is default, we will just do nothing with it. So no message will be seen over here.

**Review by ading1977 — APPROVED** [review]

**Pull request merged**